File: lists/arrayList/javascript/arrayList.js
//
// ArrayList.js
// By Scott Hall
//
class ArrayList {

  // Initialize the object.  If isStrict is true then the list only allows
  // data types consistent with the first item added
  constructor(size, isStrict) {
    if (isStrict == undefined || isStrict == false) {
      this.isStrict = false;
    } else {
      this.isStrict = true;
    }
    this.maxSize = size;
    this.length  = 0;
    this.aList   = [];
    this.aType   = undefined;
  }

  // --------------------------------------------------------------------------

  // Get the max size of the list
  get getMaxSize() {
    return this.maxSize;
  }

  // --------------------------------------------------------------------------

  // Return whether the list is full or not
  get isFull() {
    if (this.length == this.maxSize) {
      return true;
    }
    return false;
  }

  // --------------------------------------------------------------------------

  // Return whether the list is empty or not
  get isEmpty() {
    if (this.length == 0) {
      return true;
    }
    return false;
  }

  // --------------------------------------------------------------------------

  // Clear the list. This simply sets the length indexer to zero.  No actual
  // list deletion is performed.  New values will simply overwrite old values.
  clear() {
    this.length = 0;
  }

  // --------------------------------------------------------------------------

  // Check if an item is in the list
  searchArrayList(value) {
    return this.aList.indexOf(value);
  }

  // --------------------------------------------------------------------------

  // Insert an item at the end of the list, providing it does not exist
  insertAtEnd(value) {

    // Is the list already full?
    if (this.isFull) {
      console.log("insertAtEnd Error:  List is full. Cannot add new item");
      return false;
    }

    // Is item already in the list?
    if (this.searchArrayList(value) != -1) {
      console.log("insertAtEnd Error: List already contains value '%s'.  Cannot add item", value);
      return false;
    }

    // If the list is strict then check the type of the value prior to adding
    if (this.isStrict && this.isEmpty == false) {
      if (this.aType != typeof(value)) {
        console.log("insertAtEnd Error: ArrayList is set to strict.  Value '%s' does not match array list type.  Cannot add", value);
        return false;
      }
    }

    // Set the array type if not done so already and list is strict
    if (this.aType == undefined && this.isStrict) {
      this.aType = typeof(value);
    }

    // Otherwise we can add item
    this.aList.push(value);

    // Increment the length indexer
    this.length++;

    return true;

  }

  // --------------------------------------------------------------------------

  // Insert an item at the specified index.  This will insert the item at the
  // index specified and shift all values to the right of insertion point.
  // The list cannot be full and the item cannot exist currently in the list
  insertAtIndex(index, value) {

    // Is the list already full?
    if (this.isFull) {
      console.log("insertAtIndex Error:  List is full. Cannot add new item");
      return false;
    }

    // Is item already in the list?
    if (this.searchArrayList(value) != -1) {
      console.log("insertAtIndex Error: List already contains value '%s'.  Cannot add item", value);
      return false;
    }

    // Is the index in the appropriate range
    if (index < 0 || index > this.length) {
      console.log("insertAtIndex Error: Index is outside of list current range.  Cannot add item");
      return false;
    }

    // If the list is strict then check the type of the value prior to adding
    if (this.isStrict && this.isEmpty == false) {
      if (this.aType != typeof(value)) {
        console.log("insertAtIndex Error: ArrayList is set to strict.  Value '%s' does not match array list type.  Cannot add", value);
        return false;
      }
    }

    // Set the array type if not done so already and list is strict
    if (this.aType == undefined && this.isStrict) {
      this.aType = typeof(value);
    }

    // Let's add the value at index, remove 0 items and let this internal
    // function shift everything else
    this.aList.splice(index, 0, value);

    // Increment the length indexer
    this.length++;

    return true;
  }

  // --------------------------------------------------------------------------

  // Remove an item at a give index
  removeAtIndex(index) {

    // Is the index in the appropriate range
    if (index < 0 || index > this.length) {
      console.log("insertAtIndex Error: Index is outside of list current range.  Cannot add item");
      return false;
    }

    this.aList.splice(index, 1);

    this.length--;

    return true;

  }

  // --------------------------------------------------------------------------

  // Remove an item by value from the list
  remove(value) {

    var index = this.searchArrayList(value);

    // Is item already in the list?
    if (index == -1) {
      console.log("remove Error: List does not contain value '%s'.  Cannot remove item", value);
      return false;
    }

    this.aList.splice(index,1);

    this.length--;

  }

  // --------------------------------------------------------------------------

  // Return the item at the given index
  getItemAtIndex(index) {

    // Is the index in the appropriate range
    if (index < 0 || index > this.length) {
      console.log("getItemAtIndex Error: Index is outside of list current range.");
      return false;
    }

    return this.aList[index];
  }

  // --------------------------------------------------------------------------

  // Print the list content
  printList() {
    var i;
    console.log("\n Array List Content:");
    for (i=0; i<this.length; i++) {
      if (i != 0 && i % 10 == 0) {
        console.log("");
      }
      process.stdout.write(this.aList[i] + ", ");
    }
    console.log("\n\n");
  }

}

// --------------------------------------------------------------------------
// Main
// Do some testing

var numArrayList = new ArrayList(10, true);
var mixArrayList = new ArrayList(10);

numArrayList.insertAtEnd(1);
numArrayList.insertAtEnd(2);
numArrayList.insertAtEnd(3);
numArrayList.insertAtEnd(4);
numArrayList.insertAtEnd(5);
numArrayList.insertAtEnd(6);
numArrayList.insertAtEnd(7);
numArrayList.insertAtEnd(8);
numArrayList.insertAtEnd(9);
numArrayList.insertAtEnd("Bad Value");

console.log("\n NUMBER ONLY ARRAY LIST\n");

console.log(" Original List");
numArrayList.printList();

console.log("Let's remove item at index 4.  Should work");
numArrayList.removeAtIndex(4);
numArrayList.printList();

console.log("Let's remove item 9.  Should work");
numArrayList.remove(9);
numArrayList.printList();

console.log("Let's remove item 99.  Should fail since 99 isn't in the list");
numArrayList.remove(99);
numArrayList.printList();

console.log("Let's add item 99 to index 6.");
numArrayList.insertAtIndex(6, 99);
numArrayList.printList();

console.log("Let's return the item at index 2");
console.log("Item at index 2 is " + numArrayList.getItemAtIndex(2));

mixArrayList.insertAtEnd(1);
mixArrayList.insertAtEnd(2);
mixArrayList.insertAtEnd(3);
mixArrayList.insertAtEnd(4);
mixArrayList.insertAtEnd("Hello");
mixArrayList.insertAtEnd(6);
mixArrayList.insertAtEnd(7);
mixArrayList.insertAtEnd(8);
mixArrayList.insertAtEnd(9);
mixArrayList.insertAtEnd("Ok");

console.log("\n MIXED ARRAY LIST\n");

console.log(" Original List");
mixArrayList.printList();

console.log("Let's remove item at index 4.  Should work");
mixArrayList.removeAtIndex(4);
mixArrayList.printList();

console.log("Let's remove item 'Hello'.  Should work");
mixArrayList.remove("Hello");
mixArrayList.printList();

console.log("Let's remove item 'Not in here'.  Should fail since 99 isn't in the list");
mixArrayList.remove("Not in here");
mixArrayList.printList();

console.log("Let's add item 'Not in here' to index 6.");
mixArrayList.insertAtIndex(6, "Not in here");
mixArrayList.printList();

console.log("Let's return the item at index 2");
console.log("Item at index 2 is " + mixArrayList.getItemAtIndex(2));

File: lists/arrayList/ruby/ArrayList.rb
#
# arrayList.rb
# By Scott Hall
#
# Data structure for an Array List

class ArrayList

  # ---------------------------------------------------------------------------

  # Constructor for initializing the ArrayList.  The maximum size of the array
  # list is required.  If the list is to adhere to strict data types, meaning
  # only the array consists of the same data types, then set strict to true.
  # Otherwise the list will be a dynamic type list
  def initialize(maxSize, strict=false)
    @aList   = Array.new(maxSize)
    @maxSize = maxSize
    @length  = 0
    @strict  = strict
    @aType   = nil
  end

  # ---------------------------------------------------------------------------

  # Accessor method for retrieving the list maximum size
  def getMaxSize
    @maxSize
  end

  # ---------------------------------------------------------------------------

  # Check if the list is currently full
  def isFull()
    if @length == @maxSize
      return true
    else
      return false
    end
  end

  # ---------------------------------------------------------------------------

  # Check if the list is empty
  def isEmpty()
    if @length == 0
      return true
    else
      return false
    end
  end

  # ---------------------------------------------------------------------------

  # Clear the list.  This simply sets the @length aTypeattribute to zero which serves
  # to overwrite existing values in the array
  def clear()
    @length = 0
  end

  # ---------------------------------------------------------------------------

  # Does the list use strict data types
  def isStrict
    @strict
  end

  # ---------------------------------------------------------------------------

  # Search the array for the value specified value.  If not present then nil is
  # returned
  def searchArrayList(value)
    return @aList.index value
  end

  # ---------------------------------------------------------------------------

  # Insert an item at the end of the list if it is not already present and the
  # list is not full
  def insertAtEnd(value)

    # Is the list full?
    if isFull()
      puts "ERROR: insertAtEnd(): List is full cannot add additional items"
      return
    end

    # Is the value already in the list?
    if searchArrayList(value) != nil
      puts "ERROR: insertAtEnd(): List already contains '#{value}'"
      return
    end

    # If the list is set to consist of strict data types then check value
    # type as well
    if isStrict() == true

      # If the list is empty then the first item added will dictate the
      # list type
      if @aType == nil
        @aType = value.class
      else
        if value.class != @aType
          puts "ERROR: insertAtEnd(): List is set to strict data type '#{@aType}' and value passed is of type '#{value.class}'.  Cannot add new value."
          return
        end
      end
      # Otherwise the value can be added so proceed!
    end

    @aList[@length] = value
    @length += 1

  end

  # ---------------------------------------------------------------------------

  # Insert an item at the end of the list if it is not already present and the
  # list is not full
  def insertAtIndex(value, index)

    # Is the list full?
    if isFull()
      puts "ERROR: insertAtIndex(): List is full cannot add additional items"
      return
    end

    # Is the value already in the list?
    if searchArrayList(value) != nil
      puts "ERROR: insertAtIndex(): List already contains '#{value}'"
      return
    end

    # Is index valid?
    if (index < 0 || index > @length)
      puts "ERROR: insertAtIndex(): Index '#{index}' is outside range of list 0...'#{@length}'.  Cannot add value"
      return
    end

    # If the list is set to consist of strict data types then check value
    # type as well
    if isStrict() == true

      # If the list is empty then the first item added will dictate the
      # list type
      if @aType == nil
        @aType = value.class
      else
        if value.class != @aType
          puts "ERROR: insertAtIndex(): List is set to strict data type '#{@aType}' and value passed is of type '#{value.class}'.  Cannot add new value."
          return
        end
      end
      # Otherwise the value can be added so proceed!
    end

    # Use Ruby's built-in array insert method to handle this request
    @aList.insert(index, value)
    @length += 1

  end

  # ---------------------------------------------------------------------------

  # Remove the item located at the specified array index
  def removeAtIndex(index)

    # Check if index is in valid range
    if (index < 0 || index > @length)
      puts "ERROR: removeAtIndex(): Index '#{index}' is outside range of list 0...'#{@length}'.  Cannot add value"
      return
    end

    @aList.delete_at(index)
    @length -= 1

  end

  # ---------------------------------------------------------------------------

  # Remove the specified value from the list.  If it does not exist then
  # nothing is done.
  def remove(value)

    index = searchArrayList(value)

    if index == nil
      puts  "ERROR: remove(): Value '#{value}' does not exist in the list.  Cannot remove!"
      return
    end

    @aList.delete_at(index)
    @length -= 1

  end

  # ---------------------------------------------------------------------------

  # Check if a value in the list equals the value specified
  def itemAtEqual(index, value)

    # Check if index is in valid range
    if (index < 0 || index > @length)
      puts "ERROR: itemAtEqual(): Index '#{index}' is outside range of list 0...'#{@length}'.  Cannot add value"
      return
    end

    if @aList[index] == value
      return true
    end
    return false
  end

  # ---------------------------------------------------------------------------

  # Return the item at the specified index
  def getItemAtIndex(index)

    # Check if index is in valid range
    if (index < 0 || index > @length)
      puts "ERROR: itemAtEqual(): Index '#{index}' is outside range of list 0...'#{@length}'.  Cannot add value"
      return
    end

    return @aList[index]

  end

  # ---------------------------------------------------------------------------

  # Print the contents of the list
  def printList()

    puts "ARRAY LIST:\n"

    i = 0
    while i < @length
      print "#{@aList[i]}, "
      i+=1
      if i % 10 == 0 && i != 0
        puts ""
      end
    end
    puts ""

  end

end

# -----------------------------------------------------------------------------

# MAIN - Test functionality of the data structure

# Create a strict Integer array list
puts "INTEGER STRICT LIST\n"
intArrayList = ArrayList.new(10,true)

puts "Let's add '56' and '100' to the list"
intArrayList.insertAtEnd(56)
intArrayList.insertAtEnd(100)
intArrayList.printList()

puts "Let's add '99' to index 1 in the list.  Should work"
intArrayList.insertAtIndex(99,1)
intArrayList.printList()

puts "Let's try to add a non-Integer data type to the list.  Should fail"
intArrayList.insertAtEnd("DONT ADD!!!")

puts "Let's fill up the list"
intArrayList.insertAtEnd(1)
intArrayList.insertAtEnd(2)
intArrayList.insertAtEnd(3)
intArrayList.insertAtEnd(4)
intArrayList.insertAtEnd(5)
intArrayList.insertAtEnd(6)
intArrayList.insertAtEnd(7)
intArrayList.printList()

puts "Let's try to add one more item.  Should fail since list is full"
intArrayList.insertAtEnd(50)

puts "Then let's remove an item so we can make room"
intArrayList.removeAtIndex(4)
intArrayList.printList()

puts "Let's try to add one more item.  Should work since we made room"
intArrayList.insertAtEnd(50)
intArrayList.printList()

puts "Let's remove '100' from the list"
intArrayList.remove(100)
intArrayList.printList()

puts "Let's check if the value at index 0 is equal to 56"
puts intArrayList.itemAtEqual(0,56)

puts "Let's check if the value at index 0 is equal to 16"
puts intArrayList.itemAtEqual(0,16)

puts "Let's try to fetch something from the list"
puts intArrayList.getItemAtIndex(1)

# Create a non-strict list
puts "\nNON-STRICT LIST\n"
arrayList = ArrayList.new(10)
arrayList.insertAtEnd(1)
arrayList.insertAtEnd(3.14156)
arrayList.insertAtEnd("Hi There!")
arrayList.insertAtEnd(4)
arrayList.insertAtEnd(5)
arrayList.insertAtEnd(6)
arrayList.insertAtEnd("More nonsense")
arrayList.insertAtEnd(8.99923)
arrayList.insertAtEnd(7)
arrayList.insertAtEnd(8)
arrayList.printList()

puts "Let's try to insert another item into a full list"
arrayList.insertAtEnd("Uh Oh")

puts "Let's remove an item"
arrayList.remove("Hi There!")
arrayList.printList()
